File: Maths/DutchFlag.cpp
/* Dutch National Flag problem 0,1,2 colors.

O(N) time complexity; O(1) space complexity */

#include <bits/stdc++.h>
using namespace std;

void sort_vec(vector<int> arr){
    int lo = 0;
    int mid = 0;
    int hi = arr.size()-1;
    while(mid <= hi){
        switch(arr[mid]){
            case 0:
                swap(arr[lo++],arr[mid++]);
                break;
            case 1:
                mid++;
                break;
            case 2:
                swap(arr[mid],arr[hi--]);
                break;
        }
    }
    for(auto &x : arr)
        cout << x << " ";
    cout << endl;
    return;
}

int main(){
    vector<int> arr = {0,1,1,0,1,2,1,2,0,0,0,1};
    sort_vec(arr);
    return 0;
}

File: Maths/DutchFlag.py
# Dutch Flag problem
# O(n) time complexity, O(1) space complexity

def dutch_flag(arr):
	lo = 0
	mid = 0
	hi = len(arr)-1
	while(mid <= hi):
		if arr[mid]==0:
			arr[lo],arr[mid] = arr[mid],arr[lo]
			lo+=1
			mid+=1
		elif arr[mid]==1:
			mid+=1
		else:
			arr[mid],arr[hi] = arr[hi],arr[mid]
			hi-=1
	return arr

if __name__=='__main__':
	arr = [0,1,2,2,1,0,0,2,0,1,1,0]
	dutch_flag(arr)
	for x in arr:
		print(x,end=" ")
	print("\n")
